File: src/main/java/club/sk1er/patcher/tweaker/asm/forge/VertexLighterSmoothAoTransformer.java
package club.sk1er.patcher.tweaker.asm.forge;

import club.sk1er.patcher.tweaker.transform.PatcherTransformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldInsnNode;
import org.objectweb.asm.tree.InsnList;
import org.objectweb.asm.tree.InsnNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.VarInsnNode;

public class VertexLighterSmoothAoTransformer implements PatcherTransformer {
    @Override
    public String[] getClassName() {
        return new String[]{"net.minecraftforge.client.model.pipeline.VertexLighterSmoothAo"};
    }

    @Override
    public void transform(ClassNode classNode, String name) {
        for (MethodNode method : classNode.methods) {
            if (method.name.equals("updateLightmap")) {
                clearInstructions(method);
                method.instructions.insert(fasterLightmapCalculation());
                break;
            }
        }
    }

    private InsnList fasterLightmapCalculation() {
        InsnList list = new InsnList();
        list.add(new VarInsnNode(Opcodes.ALOAD, 0));
        list.add(new FieldInsnNode(Opcodes.GETFIELD, "net/minecraftforge/client/model/pipeline/VertexLighterSmoothAo", "blockInfo", "Lnet/minecraftforge/client/model/pipeline/BlockInfo;"));
        list.add(new VarInsnNode(Opcodes.ALOAD, 2));
        list.add(new VarInsnNode(Opcodes.FLOAD, 3));
        list.add(new VarInsnNode(Opcodes.FLOAD, 4));
        list.add(new VarInsnNode(Opcodes.FLOAD, 5));
        list.add(new MethodInsnNode(Opcodes.INVOKESTATIC, getHooksPackage("VertexLighterSmoothAoHook"), "calcLightmap", "(Lnet/minecraftforge/client/model/pipeline/BlockInfo;[FFFF)V", false));
        list.add(new InsnNode(Opcodes.RETURN));
        return list;
    }
}

File: src/main/kotlin/club/sk1er/patcher/asm/forge/VertexLighterSmoothAoTransformer.kt
/*
 * Copyright © 2020 by Sk1er LLC
 *
 * All rights reserved.
 *
 * Sk1er LLC
 * 444 S Fulton Ave
 * Mount Vernon, NY
 * sk1er.club
 */
package club.sk1er.patcher.asm.forge

import club.sk1er.hookinjection.injectInstructions
import club.sk1er.patcher.hooks.VertexLighterSmoothAoHook
import club.sk1er.patcher.tweaker.transform.PatcherTransformer
import codes.som.anthony.koffee.assembleBlock
import codes.som.anthony.koffee.insns.jvm.*
import org.objectweb.asm.tree.ClassNode

class VertexLighterSmoothAoTransformer : PatcherTransformer {
    /**
     * The class name that's being transformed
     *
     */
    override fun getClassName() = arrayOf("net.minecraftforge.client.model.pipeline.VertexLighterSmoothAo")

    /**
     * Perform any asm in order to transform code
     *
     * @param classNode the transformed class node
     * @param name      the transformed class name
     */
    override fun transform(classNode: ClassNode, name: String) {
        for (methodNode in classNode.methods) {
            if (methodNode.name == "calcLightmap") {
                clearInstructions(methodNode)
                methodNode.desc = "([FFFF)V"
                injectInstructions {
                    of(VertexLighterSmoothAoHook::fastCalcLightmap)
                    into(methodNode)
                    params(0, 1, 2, 3, 4)
                    keepReturns
                }
            } else if (methodNode.name == "updateLightmap") {
                clearInstructions(methodNode)
                methodNode.instructions.insert(assembleBlock {
                    aload_0
                    aload_2
                    fload_3
                    fload(4)
                    fload(5)
                    invokevirtual(
                        "net/minecraftforge/client/model/pipeline/VertexLighterSmoothAo",
                        "calcLightmap",
                        void,
                        FloatArray::class,
                        float,
                        float,
                        float
                    )
                    _return
                }.first)
            }
        }
    }
}